drivers/sensor/npm1300_charger/npm1300_charger_trigger.c

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_npm1300_charger
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/mfd/npm1300.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/sensor/npm1300_charger.h>
+
+#include "npm1300_charger.h"
+
+/* nPM1300 base addresses */
+#define MAIN_BASE 0x00U
+
+/* nPM1300 MAIN register offsets */
+#define MAIN_OFFSET_SET      0x00U
+#define MAIN_OFFSET_CLR      0x01U
+#define MAIN_OFFSET_INTENSET 0x02U
+#define MAIN_OFFSET_INTENCLR 0x03U
+
+#define MAIN_SIZE 0x26U
+
+struct event_reg_t {
+	struct sensor_trigger trig;
+	uint8_t offset;
+	uint8_t mask;
+};
+
+static const struct event_reg_t event_reg[NPM1300_TRIGGER_HANDLERS] = {
+	/* Charge completed event */
+	{{SENSOR_TRIG_THRESHOLD, SENSOR_CHAN_NPM1300_CHARGER_STATUS}, 0x0AU, 0x10U},
+	/* Charge error event */
+	{{SENSOR_TRIG_THRESHOLD, SENSOR_CHAN_NPM1300_CHARGER_ERROR}, 0x0AU, 0x20U},
+	/* Battery detected event */
+	{{SENSOR_TRIG_THRESHOLD, SENSOR_CHAN_GAUGE_VOLTAGE}, 0x0EU, 0x01U},
+	/* Battery removed event */
+	{{SENSOR_TRIG_STATIONARY, SENSOR_CHAN_GAUGE_VOLTAGE}, 0x0EU, 0x02U},
+	/* Shiphold button press event */
+	{{SENSOR_TRIG_TAP, SENSOR_CHAN_NPM1300_SHIPHOLD}, 0x12U, 0x01U},
+	/* Watchdog timer warning event */
+	{{SENSOR_TRIG_THRESHOLD, SENSOR_CHAN_NPM1300_TIMER}, 0x12U, 0x08U},
```

Comment:
hmm shouldn't we restrict to charging events here?

Btw, since this is adding support for events, it would be nice to make sure the proposed charger API will support this. Sensor triggers have always been a mess (see all the thread/workq madness duplicated everywhere), so charger API should offer something better from day 0.

Reply:
Thanks for the feedback.  I'm not really clear on the best place to put this to be honest.
For this device the trigger/irq/event stuff needs to be handled in one place as they all share a single IRQ pin.

Should it be done in the mfd driver instead, or somewhere else?

Reply:
Makes sense, maybe the mfd driver should have a single IRQ handler, and other drivers can subscribe to it (maybe like how GPIO callbacks work actually)

Reply:
Cheers, I'll do some investigation work on that approach.